lib/ret/storage.ex

```diff
@@ -61,19 +72,62 @@ defmodule Ret.Storage do
   end
 
   def fetch(id, key) when is_binary(id) and is_binary(key) do
-    fetch_blob(id, key, @expiring_file_path)
+    fetch_blob(id, key, expiring_file_path())
   end
 
   def fetch(%OwnedFile{owned_file_uuid: id, key: key}) do
-    fetch_blob(id, key, @owned_file_path)
+    fetch_blob(id, key, owned_file_path())
+  end
+
+  def fetch(%CachedFile{} = cached_file) do
+    fetch_at(cached_file, Timex.now() |> Timex.to_naive_datetime() |> NaiveDateTime.truncate(:second))
+  end
+
+  defp maybe_bump_accessed_at(%CachedFile{accessed_at: accessed_at} = cached_file, time) do
+    one_day_ago = Timex.shift(time, days: -1)
+
+    if accessed_at < one_day_ago do
+      cached_file |> Ecto.Changeset.change(accessed_at: time) |> Ret.Repo.update()
+    end
+  end
+
+  def fetch_at(%CachedFile{file_uuid: id, file_key: key} = cached_file, time) do
+    maybe_bump_accessed_at(cached_file, time)
+
+    case fetch_blob(id, key, cached_file_path()) do
+      {:ok, meta, stream} ->
+        {:ok, meta, stream}
+
+      {:error, _} ->
+        migrate(%{cached_file: cached_file})
+        fetch_blob(id, key, cached_file_path())
+    end
+  end
+
+  # TODO: Remove this code once all the CachedFiles are stored in the cached_file_path().
+  defp migrate(%{
+         cached_file: %CachedFile{file_uuid: id, file_key: file_key} = cached_file
+       }) do
+    with {:ok, _meta, _stream} <- fetch_blob(id, file_key, expiring_file_path()),
+         {:ok, uuid} <- Ecto.UUID.cast(id),
+         [_src_file_directory, src_meta_file_path, src_blob_file_path] <- paths_for_uuid(uuid, expiring_file_path()),
+         [dest_file_directory, dest_meta_file_path, dest_blob_file_path] <- paths_for_uuid(uuid, cached_file_path()),
+         {:ok, _} <- File.mkdir_p(dest_file_directory),
+         {:ok, _} <- File.cp(src_meta_file_path, dest_meta_file_path),
+         {:ok, _} <- File.cp(src_blob_file_path, dest_blob_file_path) do
+      {:ok, cached_file}
+    else
+      _ -> {:error, "Migration failed"}
+    end
   end
 
   defp fetch_blob(id, key, subpath) do
     with storage_path when is_binary(storage_path) <- module_config(:storage_path),
          {:ok, uuid} <- Ecto.UUID.cast(id),
          [_file_path, meta_file_path, blob_file_path] <- paths_for_uuid(uuid, subpath),
          [{:ok, _}, {:ok, _}] <- [File.stat(meta_file_path), File.stat(blob_file_path)],
-         meta <- File.read!(meta_file_path) |> Poison.decode!(),
+         {:ok, meta_file_data} <- File.read(meta_file_path),
+         {:ok, meta} <- Poison.decode(meta_file_data),
```

Comment:
Previously this would raise an exception related to file read failure or json decode failure, but now it will be caught by fall-through `_ ->` case below, which means the details of the error will be hidden and error handling based on `:not_found` will be misleading.

Reply:
I'd say we should either change `:not_found` to a more generic error, or revert these lines and put a `try` around whatever calls `fetch_blob`.